internal/pkg/bulk/engine.go

```diff
@@ -405,7 +406,10 @@ func (b *Bulker) Run(ctx context.Context) error {
 					Int("byteCnt", byteCnt).
 					Msg("Flush on threshold")
 
-				err = doFlush()
+				// deadline prevents bulker being blocked on flush
+				flushCtx, cancel := context.WithTimeout(ctx, defaultFlushContextTimeout)
+				defer cancel()
```

Comment:
These contexts are created in a loop, but the cancels are deffered until the function (`Bulker.Run`) returns. Should we move context creation to within the `doFlush` function to prevent this?

Reply:
I originally added it to `doFlush` but it didn't work, Fleet Server didn't come up successfully. I think `doFlush` finishes too quickly and would cancel the context.

Reply:
It doesn't work because most of the work happening is asynchronous in a goroutine, so once you unblock from acquiring the semaphore if you cancel immediately you'll exit the goroutine.

https://github.com/elastic/fleet-server/blob/1a23838f0eb64ae2748b058e3839a60afed2725d/internal/pkg/bulk/engine.go#L458-L461

There probably needs to be two separate deadlines created in doFlush():

1. On the semaphore acquire call: https://github.com/elastic/fleet-server/blob/1a23838f0eb64ae2748b058e3839a60afed2725d/internal/pkg/bulk/engine.go#L446-L449
2. Inside the goroutine doing the work https://github.com/elastic/fleet-server/blob/1a23838f0eb64ae2748b058e3839a60afed2725d/internal/pkg/bulk/engine.go#L458-L472

Reply:
tried to add deadline in doFlush to those 2 places, but it doesn't' seem to work
logs: https://platform-logging.kb.us-west2.gcp.elastic-cloud.com/app/r/s/L3Nkr